### PR TITLE
Allow partial repeat of `readdir` response

### PR DESCRIPTION
**Pull request by vladem**
## Description of change
When user application gets interrupted in a `readdir` syscall the underlying chain of `readdir` fuse requests gets reset to an offset which is considered stale by Mountpoint. In that case Mountpoint still completes the interrupted `readdir` request, but kernel partially discards the response. We already cache the last response, so we can use it to serve the request which follows the interrupt. 

Relevant issues: https://github.com/awslabs/mountpoint-s3/issues/955

## Does this change impact existing behavior?

This is not a breaking change. Previously an error was returned, now it'll be handled properly.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

### CODE REVIEW

File: mountpoint-s3/src/fs.rs [hunk 1]
```diff
@@ -1016,18 +1016,22 @@ where
 
         let readdir_handle = dir_handle.handle.lock().await;
 
-        if offset != dir_handle.offset() {
+        if offset != dir_handle.offset() && offset > -1 {
```
Comment (dannycjones) [inline mountpoint-s3/src/fs.rs:1019]: nit: this is probably a bit clearer

```suggestion
        if offset != dir_handle.offset() && offset >= 0 {
```

File: mountpoint-s3/tests/fs.rs [hunk 2]
```diff
@@ -1482,6 +1482,45 @@ async fn test_readdir_rewind_with_local_files_only() {
...
+    let first_response = ls(&fs, dir_handle, 0, max_entries).await;
+    assert!(first_response.len() == 25);
+
+    // FUSE( 11) INTERRUPT unique RequestId(10)
+    // don't really interrupt the application, but emulate the resulting fuse requests
+
+    // FUSE( 12) READDIRPLUS fh FileHandle(1), offset 1, size 4096
+    // first response was partially discarded because of an interrupt, but the returned entries should be the same for the second request
+    let second_response = ls(&fs, dir_handle, 1, max_entries).await;
+    assert_eq!(&first_response[1..], &second_response[..]);
```
Comment (dannycjones) [inline mountpoint-s3/tests/fs.rs:1508]: I'd prefer we generalize over this (`test_case(..)`) and provide some edge cases.

In particular, I'm looking for:
- `1` as you have
- `24`, i.e right at the end of the page so we know we're not doing anything wrong on the edges

**Review by dannycjones — CHANGES_REQUESTED** [review]
Nice!

Thanks for adding the new test case. I think it can be easily adapted to cover some edge cases (see comment).

Do you think this deserves a new changelog entry?

File: mountpoint-s3/src/fs.rs [hunk 3]
```diff
@@ -1016,18 +1016,22 @@ where
```
Comment (arsh) [inline mountpoint-s3/src/fs.rs:1016]: on line 1010 we have a special case; can you check if removing it does not break the tests?

Comment (vladem) [reply mountpoint-s3/src/fs.rs:1016]: I've added a test for using cached response after a rewind. It works as expected, as the cached response is updated after the first request after rewind completes.

After commenting out the block on [line 1010](https://github.com/awslabs/mountpoint-s3/blob/e2680a231511832472c23c086ecb1493ad73201e/mountpoint-s3/src/fs.rs#L1010C12-L1010C35) only rewind tests are failing, which is expected: 
```
test test_readdir_rewind_with_new_files::default_default_expects ... FAILED
test test_readdir_rewind_ordered ... FAILED
test test_readdir_repeat_response_partial::first_in_the_end_second_in_full ... ok
test test_readdir_then_open_cached ... ok
test test_readdir_repeat_response_partial::first_in_the_beginning_second_in_full ... ok
test test_readdir_repeat_response_partial::first_in_full_second_in_the_end ... ok
test test_readdir_rewind_unordered ... FAILED
test test_readdir_rewind_with_local_files_only ... FAILED
test test_readdir_vs_readdirplus ... ok
test test_readdir_rewind_with_new_files::s3filesystemconfig_s3_personality_s3personality_expressonezone_default_default_expects ... FAILED
test test_readdir_repeat_response_partial::first_in_full_second_in_the_beginning ... ok
```

**Review by dannycjones — COMMENTED** [review]
Sorry, forgot to post these comments. Just a concern on outdated comments

File: mountpoint-s3/src/fs.rs (hunk 3 above)
Comment (vladem) [reply mountpoint-s3/src/fs.rs:1016]: Decided to explicitly destroy the cache on rewind.

(After another look at it I had a suspicion that when rewind happens while cached response starts at offset 0, this response will be repeated. In an unordered case it would have lead to wrong results. It proved to be wrong, as we don't try to use cache in this case as `offset == dir_handle.offset() == 0`, but destroying the cache, while being redundant, protects us from possible bugs in future)

File: mountpoint-s3/src/fs.rs (hunk 1 above)
Comment (vladem) [reply mountpoint-s3/src/fs.rs]: made it `offset > 0` since in case `offset == 0` we do the rewind and don't want to use cache (`offset != dir_handle.offset()` will also be `false`, but still)

File: mountpoint-s3/CHANGELOG.md
Comment (dannycjones) [inline mountpoint-s3/CHANGELOG.md:1]: We have a conflict here, needs a rebase.

File: mountpoint-s3/tests/fs.rs [hunk 4]
```diff
@@ -1485,6 +1485,80 @@ async fn test_readdir_rewind_with_local_files_only() {
     assert_eq!(new_entries.len(), 3); // 1 new local file + 2 dirs (. and ..) = 3 entries
 }
 
+// Check that request with an out-of-order offset which is in bounds of previously cached response is served well.
+// This is relevant for situation when user application is interrupted in a readdir system call, which makes kernel
+// to partially discard previous response and request some entires from it again:
```
Comment (dannycjones) [inline mountpoint-s3/tests/fs.rs:1490]: typo/nit

```suggestion
// Check that request with an out-of-order offset which is in bounds of previously cached response is served well.
// This is relevant for situation when user application is interrupted in a readdir system call, which makes the
// kernel partially discard previous response and request some entries from it again:
```

**Review by dannycjones — COMMENTED** [review]
Just rebase and some nits (since we're already updating the branch).

**Comment by dannycjones** [comment]
@vladem would you mind rebasing this one?

**Review by dannycjones — APPROVED** [review]
LGTM, one minor nit I missed before

**Review by dannycjones — APPROVED** [review]
LGTM!
